File: sdk/typescript/src/actions/commentary.ts
import { jsonRequest, listParams, pathId, requestInit, withParams } from "../internal/utils.js";
import type {
  Comment,
  CommentListResponse,
  CommentMentionCandidateListResponse,
  CommentMentionCandidateParams,
  CommentResponse,
  CommentThread,
  CommentThreadListResponse,
  CommentThreadResponse,
  CreateCommentRequest,
  ListParams,
  UpdateCommentRequest,
  UUID,
} from "../types.js";
import type { KivalClientBase, WithSignal } from "./types.js";

/** Parameters for {@link listObjectCommentary}. */
export type ListObjectCommentaryParameters = WithSignal<
  ListParams & {
    workspaceId: UUID;
    objectId: UUID;
  }
>;

/** Return type for {@link listObjectCommentary}. */
export type ListObjectCommentaryReturnType = CommentThreadListResponse;

/** Lists object commentary ordered by latest thread activity. */
export function listObjectCommentary(
  client: KivalClientBase,
  parameters: ListObjectCommentaryParameters,
): Promise<ListObjectCommentaryReturnType> {
  const { workspaceId, objectId, ...params } = parameters;
  return client.requestJson<ListObjectCommentaryReturnType>(
    withParams(commentaryPath(workspaceId, objectId), listParams(params)),
    requestInit({}, params.signal),
  );
}

/** Parameters for {@link listCommentThreadComments}. */
export type ListCommentThreadCommentsParameters = WithSignal<
  ListParams & {
    workspaceId: UUID;
    objectId: UUID;
    threadId: UUID;
  }
>;

/** Return type for {@link listCommentThreadComments}. */
export type ListCommentThreadCommentsReturnType = CommentListResponse;

/** Lists a page of comments in one commentary thread. */
export function listCommentThreadComments(
  client: KivalClientBase,
  parameters: ListCommentThreadCommentsParameters,
): Promise<ListCommentThreadCommentsReturnType> {
  const { workspaceId, objectId, threadId, ...params } = parameters;
  return client.requestJson<ListCommentThreadCommentsReturnType>(
    withParams(
      `${commentaryPath(workspaceId, objectId)}/${pathId(threadId)}/comments`,
      listParams(params),
    ),
    requestInit({}, params.signal),
  );
}

/** Parameters for {@link listCommentMentionCandidates}. */
export type ListCommentMentionCandidatesParameters = WithSignal<
  CommentMentionCandidateParams & {
    workspaceId: UUID;
    objectId: UUID;
  }
>;

/** Return type for {@link listCommentMentionCandidates}. */
export type ListCommentMentionCandidatesReturnType = CommentMentionCandidateListResponse;

/** Lists active users who can view the object and may be mentioned from commentary. */
export function listCommentMentionCandidates(
  client: KivalClientBase,
  parameters: ListCommentMentionCandidatesParameters,
): Promise<ListCommentMentionCandidatesReturnType> {
  const { workspaceId, objectId, q, limit, signal } = parameters;
  const params = new URLSearchParams();
  if (q) {
    params.set("q", q);
  }
  if (limit != null) {
    params.set("limit", limit.toString());
  }

  return client.requestJson<ListCommentMentionCandidatesReturnType>(
    withParams(`${commentaryPath(workspaceId, objectId)}/mention-candidates`, params),
    requestInit({}, signal),
  );
}

/** Parameters for {@link createCommentThread}. */
export type CreateCommentThreadParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
  input: CreateCommentRequest;
}>;

/** Return type for {@link createCommentThread}. */
export type CreateCommentThreadReturnType = CommentThread;

/** Creates a top-level commentary thread. */
export function createCommentThread(
  client: KivalClientBase,
  parameters: CreateCommentThreadParameters,
): Promise<CreateCommentThreadReturnType> {
  const { workspaceId, objectId, input, signal } = parameters;
  return client
    .requestJson<CommentThreadResponse>(
      commentaryPath(workspaceId, objectId),
      jsonRequest("POST", input, signal),
    )
    .then((response) => response.thread);
}

/** Parameters for {@link replyToCommentThread}. */
export type ReplyToCommentThreadParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
  threadId: UUID;
  input: CreateCommentRequest;
}>;

/** Return type for {@link replyToCommentThread}. */
export type ReplyToCommentThreadReturnType = Comment;

/** Replies to an open commentary thread. */
export function replyToCommentThread(
  client: KivalClientBase,
  parameters: ReplyToCommentThreadParameters,
): Promise<ReplyToCommentThreadReturnType> {
  const { workspaceId, objectId, threadId, input, signal } = parameters;
  return client
    .requestJson<CommentResponse>(
      `${commentaryPath(workspaceId, objectId)}/${pathId(threadId)}/replies`,
      jsonRequest("POST", input, signal),
    )
    .then((response) => response.comment);
}

/** Parameters for {@link updateComment}. */
export type UpdateCommentParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
  commentId: UUID;
  input: UpdateCommentRequest;
}>;

/** Return type for {@link updateComment}. */
export type UpdateCommentReturnType = Comment;

/** Edits an active comment authored by the current user in an open thread. */
export function updateComment(
  client: KivalClientBase,
  parameters: UpdateCommentParameters,
): Promise<UpdateCommentReturnType> {
  const { workspaceId, objectId, commentId, input, signal } = parameters;
  return client
    .requestJson<CommentResponse>(
      `${commentaryPath(workspaceId, objectId)}/comments/${pathId(commentId)}`,
      jsonRequest("PATCH", input, signal),
    )
    .then((response) => response.comment);
}

/** Parameters for {@link deleteComment}. */
export type DeleteCommentParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
  commentId: UUID;
}>;

/** Return type for {@link deleteComment}. */
export type DeleteCommentReturnType = Comment;

/** Soft-deletes a comment as its author or an object administrator. */
export function deleteComment(
  client: KivalClientBase,
  parameters: DeleteCommentParameters,
): Promise<DeleteCommentReturnType> {
  const { workspaceId, objectId, commentId, signal } = parameters;
  return client
    .requestJson<CommentResponse>(
      `${commentaryPath(workspaceId, objectId)}/comments/${pathId(commentId)}`,
      requestInit({ method: "DELETE" }, signal),
    )
    .then((response) => response.comment);
}

/** Parameters for {@link resolveCommentThread}. */
export type ResolveCommentThreadParameters = WithSignal<{
  workspaceId: UUID;
  objectId: UUID;
  threadId: UUID;
}>;

/** Return type for {@link resolveCommentThread}. */
export type ResolveCommentThreadReturnType = CommentThread;

/** Resolves an open commentary thread. */
export function resolveCommentThread(
  client: KivalClientBase,
  parameters: ResolveCommentThreadParameters,
): Promise<ResolveCommentThreadReturnType> {
  return setCommentThreadResolution(client, parameters, "resolve");
}

/** Parameters for {@link reopenCommentThread}. */
export type ReopenCommentThreadParameters = ResolveCommentThreadParameters;

/** Return type for {@link reopenCommentThread}. */
export type ReopenCommentThreadReturnType = CommentThread;

/** Reopens a resolved commentary thread. */
export function reopenCommentThread(
  client: KivalClientBase,
  parameters: ReopenCommentThreadParameters,
): Promise<ReopenCommentThreadReturnType> {
  return setCommentThreadResolution(client, parameters, "reopen");
}

function setCommentThreadResolution(
  client: KivalClientBase,
  parameters: ResolveCommentThreadParameters,
  action: "resolve" | "reopen",
): Promise<CommentThread> {
  const { workspaceId, objectId, threadId, signal } = parameters;
  return client
    .requestJson<CommentThreadResponse>(
      `${commentaryPath(workspaceId, objectId)}/${pathId(threadId)}/${action}`,
      requestInit({ method: "POST" }, signal),
    )
    .then((response) => response.thread);
}

function commentaryPath(workspaceId: UUID, objectId: UUID) {
  return `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/commentary`;
}

File: sdk/rust/src/client/commentary.rs
//! Object commentary API operations.

use reqwest::Method;
use uuid::Uuid;

use crate::{
    ClientError, Comment, CommentListResponse, CommentMentionCandidateListResponse,
    CommentMentionCandidateParams, CommentResponse, CommentThread, CommentThreadListResponse,
    CommentThreadResponse, CreateCommentRequest, KivalClient, ListParams, Transport,
    UpdateCommentRequest, client::transport::append_list_params,
};

impl<S> KivalClient<S>
where
    S: Transport,
{
    /// Lists commentary attached to an object.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is configured or the request fails.
    pub async fn list_object_commentary(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        params: &ListParams,
    ) -> Result<CommentThreadListResponse, ClientError> {
        self.require_api_key()?;

        let mut url =
            self.api_url(&format!("/workspaces/{workspace_id}/objects/{object_id}/commentary"))?;
        append_list_params(&mut url, params);

        let request = self.request_url(&Method::GET, url)?;
        self.send_json(request).await
    }

    /// Lists a page of comments in one commentary thread.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is configured or the request fails.
    pub async fn list_comment_thread_comments(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        thread_id: Uuid,
        params: &ListParams,
    ) -> Result<CommentListResponse, ClientError> {
        self.require_api_key()?;

        let mut url = self.api_url(&format!(
            "/workspaces/{workspace_id}/objects/{object_id}/commentary/{thread_id}/comments"
        ))?;
        append_list_params(&mut url, params);

        let request = self.request_url(&Method::GET, url)?;
        self.send_json(request).await
    }

    /// Lists active users who can currently view the object and may be mentioned.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is configured or the request fails.
    pub async fn list_comment_mention_candidates(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        params: &CommentMentionCandidateParams,
    ) -> Result<CommentMentionCandidateListResponse, ClientError> {
        self.require_api_key()?;

        let mut url = self.api_url(&format!(
            "/workspaces/{workspace_id}/objects/{object_id}/commentary/mention-candidates"
        ))?;
        {
            let mut pairs = url.query_pairs_mut();
            if !params.q.is_empty() {
                pairs.append_pair("q", &params.q);
            }
            if let Some(limit) = params.limit {
                pairs.append_pair("limit", &limit.to_string());
            }
        }

        let request = self.request_url(&Method::GET, url)?;
        self.send_json(request).await
    }

    /// Creates a top-level commentary thread.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is configured or the request fails.
    pub async fn create_comment_thread(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        request: &CreateCommentRequest,
    ) -> Result<CommentThread, ClientError> {
        self.require_api_key()?;

        let request = self
            .request(
                &Method::POST,
                &format!("/workspaces/{workspace_id}/objects/{object_id}/commentary"),
            )?
            .json(request);
        let response = self.send_json::<CommentThreadResponse>(request).await?;
        Ok(response.thread)
    }

    /// Replies to an open commentary thread.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is configured or the request fails.
    pub async fn reply_to_comment_thread(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        thread_id: Uuid,
        request: &CreateCommentRequest,
    ) -> Result<Comment, ClientError> {
        self.require_api_key()?;

        let request = self
            .request(
                &Method::POST,
                &format!(
                    "/workspaces/{workspace_id}/objects/{object_id}/commentary/{thread_id}/replies"
                ),
            )?
            .json(request);
        let response = self.send_json::<CommentResponse>(request).await?;
        Ok(response.comment)
    }

    /// Edits a comment authored by the current user in an open thread.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is configured or the request fails.
    pub async fn update_comment(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        comment_id: Uuid,
        request: &UpdateCommentRequest,
    ) -> Result<Comment, ClientError> {
        self.require_api_key()?;

        let request = self
            .request(
                &Method::PATCH,
                &format!(
                    "/workspaces/{workspace_id}/objects/{object_id}/commentary/comments/{comment_id}"
                ),
            )?
            .json(request);
        let response = self.send_json::<CommentResponse>(request).await?;
        Ok(response.comment)
    }

    /// Soft-deletes a comment authored by the current user or moderated by an object admin.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is configured or the request fails.
    pub async fn delete_comment(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        comment_id: Uuid,
    ) -> Result<Comment, ClientError> {
        self.require_api_key()?;

        let request = self.request(
            &Method::DELETE,
            &format!(
                "/workspaces/{workspace_id}/objects/{object_id}/commentary/comments/{comment_id}"
            ),
        )?;
        let response = self.send_json::<CommentResponse>(request).await?;
        Ok(response.comment)
    }

    /// Resolves an open commentary thread.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is configured or the request fails.
    pub async fn resolve_comment_thread(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        thread_id: Uuid,
    ) -> Result<CommentThread, ClientError> {
        self.set_comment_thread_resolution(workspace_id, object_id, thread_id, "resolve").await
    }

    /// Reopens a resolved commentary thread.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is configured or the request fails.
    pub async fn reopen_comment_thread(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        thread_id: Uuid,
    ) -> Result<CommentThread, ClientError> {
        self.set_comment_thread_resolution(workspace_id, object_id, thread_id, "reopen").await
    }

    /// Sets the resolution state of a commentary thread using the given API action.
    ///
    /// # Errors
    ///
    /// Returns an error if no API key is configured or the request fails.
    async fn set_comment_thread_resolution(
        &self,
        workspace_id: Uuid,
        object_id: Uuid,
        thread_id: Uuid,
        action: &str,
    ) -> Result<CommentThread, ClientError> {
        self.require_api_key()?;

        let request = self.request(
            &Method::POST,
            &format!(
                "/workspaces/{workspace_id}/objects/{object_id}/commentary/{thread_id}/{action}"
            ),
        )?;
        let response = self.send_json::<CommentThreadResponse>(request).await?;
        Ok(response.thread)
    }
}
